File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.py
# 맥주 1 박스 = 20 병
# 50미터 = 1 병
# 편의점 = 20 병

from collections import deque

t = int(input())

def bfs():
    q = deque([(home_x, home_y)])
    
    while q:
        x, y = q.popleft()
        if abs(x - kanye_x) + abs(y - kanye_y) <= 1000:
            print('happy')
            return
        for i in range(n):
            if not visited[i]:
                nx, ny = gs25[i]
                if abs(x-nx) + abs(y-ny) <= 1000:
                    visited[i] = 1
                    q.append([nx, ny])

    print('sad')
                

for _ in range(t):
    n = int(input())
    home_x, home_y = map(int, input().split())
    gs25 = [
        list(map(int, input().split()))
        for _ in range(n)
    ]
    kanye_x, kanye_y = map(int, input().split())
    visited = [0] * (n+1)
    bfs()

File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.java
import java.io.*;
import java.util.*;

public class Main {

    static int N;
    static ArrayList<Point> home;
    static ArrayList<Point> cu;
    static ArrayList<Point> rock;
    static boolean[] v;

    static class Point {
        int i, j;

        Point(int i, int j) {
            this.i = i;
            this.j = j;
        }
    }

    static boolean bfs() {
        Deque<int[]> q = new ArrayDeque<>();
        q.offer(new int[] {home.get(0).i, home.get(0).j});

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int ci = cur[0];
            int cj = cur[1];

            if ((Math.abs(ci - rock.get(0).i) + Math.abs(cj - rock.get(0).j)) <= 1000) return true;

            for (int f = 0; f < N; f++) {
                if (!v[f]) {
                    int ni = cu.get(f).i;
                    int nj = cu.get(f).j;
                    if ((Math.abs(ni - ci) + Math.abs(nj - cj)) <= 1000) {
                        q.offer(new int[]{ni, nj});
                        v[f] = true;
                    }
                }
            }
        }
        return false;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int T = Integer.parseInt(st.nextToken());
        for (int tc = 1; tc <= T; tc++) {
            N = Integer.parseInt(br.readLine());
            home = new ArrayList<>();
            cu = new ArrayList<>();
            rock = new ArrayList<>();
            st = new StringTokenizer(br.readLine());
            home.add(new Point(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())));
            for (int i = 0; i < N; i++) {
                st = new StringTokenizer(br.readLine());
                cu.add(new Point(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())));
            }
            st = new StringTokenizer(br.readLine());
            rock.add(new Point(Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())));
            v = new boolean[N];

            String ans = "";
            if (bfs()) ans = "happy";
            else ans = "sad";

            System.out.println(ans);
        }
    }
}